druid-shell/src/platform/windows/window.rs

```diff
@@ -357,6 +370,57 @@ impl MyWndProc {
     fn has_menu(&self) -> bool {
         self.with_window_state(|state| state.has_menu.get())
     }
+
+    // Here we handle messages generated by WindowHandle
+    // that needs to run after borrow is released
+    fn handle_blocked(&self) {
```

Comment:
This is problematic for `open_file_sync()` and `save_as_sync()` as they need to return `Option<FileInfo>`

Any idea how we could handle `get_file_dialog_path()` here but get the return value in WindowHandle?

Reply:
Rather than add a callback, I'd suggest adding methods to the `WinHandler` trait, one for every blocking operation that needs a "reply".

Reply:
Should it then trigger a Event? That would work for widgets, but not for button closure e.g:
```
let button = Button::new("Select File").on_click(|ctx, _data: &mut String, _env| {
        let i = ctx
            .window()
            .clone()
            .open_file_sync(FileDialogOptions::new());
    });
```

Reply:
Hm, I might have changed my mind: a callback has certain advantages. But on the other hand, the two-call protocol is being introduced in #1039, and I think it's important to be consistent. @xStrom?

The way the two-call protocol would work, is that druid would call `WindowHandle::open_file_dialog` (or some name like that) and then when the file dialog is closed, druid-shell will call `WinHandler::file_dialog_finished` (or some name like that). Druid's implementation of `WinHandler` will take care of sending the `OPEN_FILE` (or `OPEN_PANEL_CANCELLED`, once #1061 lands) command down the widget tree.

Then we would deprecate `open_file_sync`. Users of druid should open files by sending a `SHOW_OPEN_PANEL` command and then listening for a `OPEN_FILE` command.

Reply:
Sorry for the spam; I just realized the problem with a callback: it's hard to do anything useful in a callback, because most of the useful things (like modifying the data, or sending a command) won't pass the borrow checker.

Reply:
Agreed, We can hold of on changing `open_file_sync()` and `save_as_sync()` for later, does not have to be in this PR.